### PR TITLE
Expose flip operations from the display matrix

### LINKED ISSUE #317: Missing  vertical/horizontal flip
since Oct 18, 2016, with https://github.com/FFMS/ffms2/commit/2ed27a5067aa2e8aedb809f19ac7941d770775dd
the rotation is exposed, but does not include flip.

For flip, you are supposed to look at the determinant of the matrix:
determinant = matrix[0,0] * matrix[1,1] - matrix[0.1]*matrix[1,0]

If the determinant is negative, we have a horizontal flip and we are suppose to apply a horizontal flip to the matrix from left before getting the rotation angle. 

also, if the rotation is 180 and the determinant is negative, we should just apply an vertical flip and not a rotation of 180

Will it not be better to expose the whole matrix and let user to decide how to interpret it, instead of doing it in the code?

### PR DESCRIPTION
**Pull request by kodabb**
Fix #317.

### CODE REVIEW

**Comment by kodabb** [comment]
some notes:
* the `VP.Rotation` values do not change if the video is not v/h flipped
* they do change if it is, but the code was doing the wrong thing anyway -- applications need to check the `VP.Flip` value
* is it worth to introduce some enums for the flip values instead of positive/negative states?
* should i add unit tests using the samples from the bug?

**Comment by lotharkript** [comment]
can we not have the full matrix instead of all those fields?

**Comment by kodabb** [comment]
@lotharkript I thought that there was consensus last year to avoid exposing it and keep the ffms2 api as simple as possible. Also `Flip` is the last field to be needed, given that all the other operations from the matrix (rotation, scale, shear) are either supported differently or do not apply to a video surface

File: src/core/videosource.cpp [hunk 1]
```diff
@@ -285,12 +285,29 @@ FFMS_VideoSource::FFMS_VideoSource(const char *SourceFile, FFMS_Index &Index, in
...
             int rot = lround(av_display_rotation_get(RotationMatrix));
             if (rot < 0)
                 rot += 360;
-            VP.Rotation = rot;
+
+            if (rot == 180 && det < 0) {
+                /* This is a vertical flip with no rotation. */
+                VP.Flip = -1;
+            } else {
+                /* It is possible to have a 90º/270º rotation and a horizontal flip. */
```
Comment (dwbuiten) [inline src/core/videosource.cpp]: Keep comments ASCII please.

**Comment by dwbuiten** [comment]
> * the `VP.Rotation` values do not change if the video is not v/h flipped
> * they do change if it is, but the code was doing the wrong thing anyway -- applications need to check the `VP.Flip` value

It's not clear from the API or documentation what will change in their values and/or what order to apply things, to me.

> is it worth to introduce some enums for the flip values instead of positive/negative states?

IMO, no.

> should i add unit tests using the samples from the bug?

They can probably be covered pretty easily, possibly within the existing tests (?)

> I thought that there was consensus last year to avoid exposing it and keep the ffms2 api as simple as possible. Also `Flip` is the last field to be needed, given that all the other operations from the matrix (rotation, scale, shear) are either supported differently or do not apply to a video surface.

Are we sure we don't need to care about things like shear/scale/etc? As in, What Would QuickTime Do? (WWQTD).

Are there any other benefits to exposing the matrix directly? That could get pretty ugly for non-mp4/mov formats, as you'd have to expose that info differently, or build of a matrix to export.

**Comment by kodabb** [comment]
> It's not clear from the API or documentation what will change in their values and/or what order to apply things, to me.

The change is that we flip the matrix *before* we compute the rotation angle. So if you had hflipped video,  before you had a rotation of 180 while now you have `rotation = 0` and `flip = 1`. It's technically an abi break *but* it only applies to videos that were previously unsupported and which we we were reporting false information of.

> They can probably be covered pretty easily, possibly within the existing tests (?)

yeah i can just copy the matrix values from the test files, but the sample ones are pretty small anyway

> Are we sure we don't need to care about things like shear/scale/etc? As in, What Would QuickTime Do? (WWQTD).
>
> Are there any other benefits to exposing the matrix directly? That could get pretty ugly for non-mp4/mov formats, as you'd have to expose that info differently, or build of a matrix to export.

I'll let @lotharkript answer this :)

File: src/core/videosource.cpp (hunk 1 above)
Comment (kodabb) [reply src/core/videosource.cpp]: done

**Comment by kodabb** [comment]
updated the comment, waiting on instructions on what to do with tests and the matrix :)

**Comment by kodabb** [comment]
I slightly modified the logic for how rotation is handled -- from the commit log

```    It should be noted that previously the rotation value was used as is
    and it was modified to be always positive by adding 360. However since
    flip operations are exposed now, it is necessary to apply some more
    logic and internally convert the rotation angle.
    
    For non-flipped videos this should be a transparent change.```

**Comment by dwbuiten** [comment]
I'll wait for a comment from @lotharkript on the whole matrix thing, unless he has no strong feelings on it.

**Comment by dwbuiten** [comment]
Having not heard from him, I'll merge once a unit test is added.

File: test/Makefile [hunk 2]
```diff
@@ -26,10 +26,19 @@ CXXFLAGS += -g -Wall -Wextra -pthread -std=c++11 -fvisibility=hidden
 
 # All tests produced by this Makefile. Remember to add new tests you
 # created to the list.
-TESTS = indexer hdr
+TESTS = indexer        \
+        hdr            \
+        display_matrix
 
+TESTS = display_matrix
```
Comment (dwbuiten) [inline test/Makefile]: You added this line by mistake, and nuked the rest of the tests.

Comment (kodabb) [reply test/Makefile]: whoops yes i was skipping the other ones, updating...

**Pull request merged**